object_intern.go

```diff
@@ -202,14 +207,18 @@ func (oi *ObjectIntern) AddOrGet(obj []byte, safe bool) (uintptr, error) {
 
 	// if neither of those terms is true then we can avoid costly allocations
 	// acquire lock
-	oi.Lock()
+	oi.RLock()
 
 	addr, ok := oi.getAndIncrement(obj)
 	if ok {
-		oi.Unlock()
+		oi.RUnlock()
 		return addr, nil
 	}
 
+	oi.RUnlock()
+
+	oi.Lock()
```

Comment:
same everywhere where we switch to the write lock based on a previously checked condition